### DISCUSSION TITLE
Specify the amount and multiplicity of the ports for each operations.

### DISCUSSION THREAD

**Discussion by aborgna-q** [Quantinuum/hugr#159] [Specification]
With multiports we can now define the exact number of ports each operation will have. There's a number of issues that are not currently specified in the spec:

- We should clarify the order of dataflow/constant/other edges (#23)
- We should specify which ports must have exactly one connection vs which can have any number of edges (including none).

In a related note, having a fixed number of ports means that the port counts in the node serialization are redundant. Should we drop them from the serialization specification?

**Comment by cqc-alec** [comment]
I propose that we keep ports as simple as possible at the spec level, and keep multiports as an implementation detail:

- Edges in the hugr may either have no ports associated with them, or exactly one port at each end.
- Ports always exist on Value and Static edges; never on Hierarchy, Order or Control Flow edges.
- Every port has an associated `SimpleType`. Joined ports have the same `SimpleType`.
- A `ClassicType` outgoing port may have any number of edges associated with it (including zero). A `LinearType` port, and an incoming port of any type, must have exactly one edge associated with it.

**Reply by ss2165** [reply]
The spec in my head is aligned with this, what is misaligned? Maybe everything here is not stated  explicitly?

**Reply by cqc-alec** [reply]
Yeah I agree it is aligned, though maybe not fully explicit. This was in response to:

> We should specify which ports must have exactly one connection vs which can have any number of edges (including none).

**Reply by ss2165** [reply]
Yes I think having a section early on (in the Edges section) that enumerates the points you list above would be good, even if strictly redundant.

**Comment by ss2165** [comment]
I'm in favour of dropping port counts in serialization, as long as we are happy with the constraint:
"All op **implmentations**, current and future, must be able to report the number of ports they require (for edges of every kind except Hierarchy)."

As far as I can tell that constraint is not too onerous?
